lhotse/features.py

```diff
@@ -0,0 +1,116 @@
+"""
+This file is just a rough sketch for now.
+"""
+from dataclasses import dataclass
+from typing import Union, List, Optional, Iterable, Any
+
+import numpy as np
+
+from lhotse.audio import AudioSet
+from lhotse.supervision import SupervisionSet
+from lhotse.utils import Seconds
+
+
+@dataclass
+class Features:
+    """Represents features extracted for some particular time range in a given recording and channel."""
+    recording_id: str
+    channel_id: int
+    start: Seconds
+    duration: Seconds
+
+    # Variant A: path to a single-segment lilcom archive serialized in some storage
+    storage_path: str
+
+    # Variant B: path to a multi-segment lilcom archive serialized in some storage AND key to access it
+    storage_path: str
+    storage_key: str
+
+    # When created from SupervisionSet maybe it makes sense to remember where this comes from... or not...
+    # if we limited ourselves to just using SupervisionSet, it could serve as the "storage_key"
+    supervision_id: Optional[str] = None
+
+    def load(self) -> np.ndarray:
+        pass
+
+
+@dataclass
+class FeatureSet:
+    """
+    Represents a feature manifest, and allows to read features for given recordings
+    within particular channels and time ranges.
+    When a given recording/time-range/channel is unavailable, raises a KeyError.
+    """
+    # TODO(pzelasko): we might need some efficient indexing structure,
+    #                 e.g. Dict[<recording-id>, Dict[<channel-id>, IntervalTree]] (pip install intervaltree)
+    features: List[Features]
+
+    @staticmethod
+    def from_yaml(param):
+        pass
+
+    @staticmethod
+    def to_yaml(param):
+        pass
+
+    def load(
+            self,
+            recording_id: str,
+            channel_id: int,
+            start: Seconds,
+            duration: Seconds,
+    ) -> np.ndarray:
+        # raise a KeyError when any of the requirements is not met
+        pass
+
+    def __iter__(self) -> Iterable[Features]:
+        return iter(self.features)
+
+    def __len__(self) -> int:
+        return len(self.features)
+
+
+@dataclass
+class FeatureSegment:
+    """
+    PZ: Now that I've written it, seems redundant with SupervisionSegment... maybe it should be a base class for it
+    """
+    recording_id: str
+    channel: int
+    start: Seconds
+    duration: Seconds
+
+
+class FeatureExtractor:
+    def with_audio_set(self, audio_set: AudioSet) -> 'FeatureExtractor':
```

Comment:
I'm not so sure about this FeatureExtractor thing.  It looks to me like something that maybe should be a script, or a collection of scripts, rather than a datatype.

Reply:
It's just a very rough sketch that helped me put my thoughts in code. My personal experience is that it's often convenient when you can have it both ways, with CLI being just a wrapper for the class, because then you can import that in a jupyter notebook or a one-off script and have the same functionality. But I won't insist - let's see how it plays out?